### PR TITLE
Interpolate time from runinfo timestep

### PR DESCRIPTION
**Pull request by g-kimbell**
Edit 25.06.25: Not using step.xml anymore. There is a missing column in runinfo for the time steps which we can use to interpolate.

This may solve incorrect time interpolation (#4 and https://github.com/Solid-Energy-Systems/NewareNDA/issues/104).

It looks like it works on my data and the .ndax files in the tests folder, but it would be useful to test more to find edge cases.

Method
- Read Step.xml, make a mapping of step index : time record step
- Get step index from step.nda
- Add the time record steps to time, round down to nearest whole step
- Fill timestamps based on new time values (to include any that are rounded down)
- Linear interpolate whatever did not work with the same method as before

Example result
```
Orig data   Real val   New interp   Old interp
    510.0      510.0        510.0        510.0
      NaN      520.0        520.0        520.0
      NaN      530.0        530.0        530.0
      NaN      540.0        540.0        540.0
      NaN      550.0        550.0        550.0
      NaN      560.0        560.0        560.0
      NaN      570.0        570.0        570.0
    580.0      580.0        580.0        580.0
      NaN      590.0        590.0        588.0  | incorrectly interpolates
      NaN      600.0        600.0        596.1  | when other measurements
      NaN      610.0        610.0        604.1  | triggers happen
      NaN      610.0        620.0        612.2  |
    620.3      620.2        620.2        620.2  < voltage diff triggers measurement
    630.0      630.0        630.0        630.0
    640.0      640.0        640.0        640.0
      NaN      650.0        650.0        650.0
      NaN      660.0        660.0        660.0
      NaN      670.0        670.0        670.0
      NaN      680.0        680.0        680.0
      NaN      690.0        690.0        690.0
      NaN      700.0        700.0        700.0
    710.0      710.0        710.0        710.0
```

When you have multiple time record conditions, both trigger. E.g. if I have a global record 10 seconds and a step record 3 seconds, it will record at 0, 3, 6, 9, 10, 12, 15, 18, 20, 21 etc.

Luckily it looks like it registers the time when it changes between the two conditions. Only adding the smaller step seems to work:

```
Orig data   Real val   New interp   Old interp
      0.0        0.0          0.0     0.000000
      3.0        3.0          3.0     3.000000
      NaN        6.0          6.0     8.333333
      NaN        9.0          9.0    13.000000
     10.0       10.0         10.0    10.000000
     12.0       12.0         12.0    12.000000
     15.0       15.0         15.0    15.000000
      NaN       18.0         18.0    20.500000
     20.0       20.0         20.0    20.000000
     21.0       21.0         21.0    21.000000
     24.0       24.0         24.0    24.000000
      NaN       27.0         27.0    29.666666
      NaN       30.0         30.0    35.000000
      NaN       33.0         33.0    40.333332
      NaN       36.0         36.0    45.666668
      NaN       39.0         39.0    51.000000
     40.0       40.0         40.0    40.000000
     42.0       42.0         42.0    42.000000
     45.0       45.0         45.0    45.000000
      NaN       48.0         48.0    50.500000
     50.0       50.0         50.0    50.000000
     51.0       51.0         51.0    51.000000
     54.0       54.0         54.0    54.000000
      NaN       57.0         57.0    59.666668
```

### CODE REVIEW

**Comment by g-kimbell** [comment]
I have discovered a edge case with this approach:

If you put e.g. "Record every 1.234 seconds" then Neware (at least my system) actually rounds up and records every 1.3 seconds, but "1234" remains as the record time in step.xml, so errors accumulate in the interpolated time.

I can add a step that rounds up to the nearest 100 ms when reading the step time, but I'm not sure if this is universal on all Neware machines.

**Comment by g-kimbell** [comment]
Another problem with the flooring step + float point errors. I round to the nearest timestep with `t = t // dt * dt` but this fails because in Python `1.3*5 // 1.3 == 4`.

I can fix this by rounding first then doing `np.floor` instead

**Comment by g-kimbell** [comment]
Several of the tests are failing now, because of 3 reasons:

1. There is a new 'Step_Index' column from read_ndax. We could drop it after interpolation, but I think it could be useful info to keep.
2. Sometimes the Time is different to the reference dataframes with the new interpolation.
3. Sometimes the "Step_Index" values seem wrong, which makes everything else go wrong. For my data (NDC version: 14 filetype: 7) it looks like Step_Index works and returns [1,2,34,5,6,7,5,6,7,5,6,7 etc.] but for example in the "test_2.ndax" (NDC version: 17 filetype: 7) the step index it gives is [1, 62, 11019, 11081 etc.] which I think is wrong. Probably the bytes are being parsed incorrectly in read_ndc.

**Comment by g-kimbell** [comment]
I had a look at NDC version: 17 filetype: 7 but could not figure out how to read the "Step_Index".

For now it skips this method if the Step_Index has values >10000 which is not robust

I also need to do more testing for what happens if the user changes the recording steps after the measurement has started with 'reset step'

**Comment by Grimler91** [comment]
It does not really change anything for the dataset I experimented with in https://github.com/d-cogswell/NewareNDA/issues/4. It changes peak sizes a bit, but they are still there

File: NewareNDA/NewareNDAx.py [hunk 1]
```diff
@@ -1,609 +1,672 @@
-# © 2022-2024 Copyright SES AI
-# Author: Daniel Cogswell
-# Email: danielcogswell@ses.ai
-
```
[the remark this answers is not in the document]
Comment (d-cogswell) [reply NewareNDA/NewareNDAx.py:4]: Yes, I was also confused about why every line changed. I just merged #10 so I also recommend changing the base for this to development.

**Comment by g-kimbell** [comment]
I was also confused why the diff went crazy. I had `git config core.autocrlf true` which converts to linux-style. I have changed it back now.

**Comment by g-kimbell** [comment]
I'm starting to wonder if this is even possible from the .ndax...
In the following example the excel export from BTS8 shows one time step record 6 seconds after instead of the usual 10 seconds.
There's no way to interpolate this or get it from the step.xml.
I guess the excel is not being converted from ndax, rather both are being exported from the SQL database in different ways?
```
Interp time  Excel reference  Orig time                    Orig timestamp  Not equal
      640.0            640.0      640.0  2025-06-02 11:35:09.783000+02:00      False
      650.0            650.0        NaN                               NaT      False
      660.0            660.0        NaN                               NaT      False
      670.0            670.0        NaN                               NaT      False
      680.0            680.0        NaN                               NaT      False
      690.0            690.0        NaN                               NaT      False
      700.0            700.0        NaN                               NaT      False
      710.0            710.0      710.0  2025-06-02 11:36:19.783000+02:00      False
      720.0            720.0        NaN                               NaT      False
      730.0            730.0        NaN                               NaT      False
      740.0            736.0        NaN                               NaT       True  <--- We cannot get this from .ndax or step.xml
      740.0            740.0      740.0  2025-06-02 11:36:49.783000+02:00      False
      750.0            750.0      750.0  2025-06-02 11:36:59.783000+02:00      False
      760.0            760.0        NaN                               NaT      False
      770.0            770.0        NaN                               NaT      False
      780.0            780.0        NaN                               NaT      False
      790.0            790.0        NaN                               NaT      False
      800.0            800.0        NaN                               NaT      False
      810.0            810.0        NaN                               NaT      False
      820.0            820.0      820.0  2025-06-02 11:38:09.783000+02:00      False
```

**Comment by g-kimbell** [comment]
I have updated the unpacking for read_ndc of the run info.
I think we were missing some of the rows, leading to missing data and impossible interpolation.

**Comment by g-kimbell** [comment]
I think I have made some progress:
- We were **missing a timestep column** in the runinfo ndc (filetype 18)
- You can forward fill and cumulative sum those timesteps, then add to Time to get the final column
- I've removed everything from step.xml, so we don't have any problems with 'Reset step'

This works great for my data, with only floating point errors checked against the raw data from Neware's 'download data' API.

This **only works for ndc14 filetype18** right now. The test data includes ndc17, which I can have a look at later. I don't have any data to play with for ndc11. For files without the timestep column it will fall back to linear interpolation.
